river/LayoutDemand.zig

```diff
@@ -0,0 +1,138 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2020 - 2021 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const std = @import("std");
+const wlr = @import("wlroots");
+const wayland = @import("wayland");
+const wl = wayland.server.wl;
+const zriver = wayland.server.zriver;
+
+const util = @import("util.zig");
+
+const Layout = @import("Layout.zig");
+const Box = @import("Box.zig");
+const Server = @import("Server.zig");
+const Output = @import("Output.zig");
+const View = @import("View.zig");
+const ViewStack = @import("view_stack.zig").ViewStack;
+
+const log = std.log.scoped(.layout);
+
+const Error = error{ViewDimensionMismatch};
+
+/// Arbitrary timeout
+const timeout = 1000;
+
+serial: u32,
+/// Number of views for which dimensions have not been pushed.
+/// This will go negative if the client pushes too many dimensions.
+views: i32,
+/// Timeout timer
+timer: *wl.EventSource,
+/// Proposed view dimensions
+view_boxen: []Box,
+
+pub fn init(layout: *Layout, views: u32) !Self {
+    // Attach a timout timer
+    const event_loop = layout.output.root.server.wl_server.getEventLoop();
+    const timer = try event_loop.addTimer(*Layout, handleTimeout, layout);
+    errdefer timer.remove();
+    try timer.timerUpdate(timeout);
+
+    return Self{
+        .serial = layout.output.root.server.wl_server.nextSerial(),
+        .views = @intCast(i32, views),
+        .timer = timer,
+        .view_boxen = try util.gpa.alloc(Box, views),
+    };
+}
+
+pub fn deinit(self: *const Self) void {
+    self.timer.remove();
+    util.gpa.free(self.view_boxen);
+}
+
+/// Destroy the LayoutDemand on timeout.
+/// All further responses to the event will simply be ignored.
+fn handleTimeout(layout: *Layout) callconv(.C) c_int {
+    log.notice(
+        "layout demand for layout '{}' on output '{}' timed out",
+        .{ layout.namespace, layout.output.wlr_output.name },
+    );
+    layout.output.layout_demand.?.deinit();
```

Comment:
Are you sure that `layout_demand` is non-null at this point?

Reply:
yep